lib/verification-reminders.js

```diff
@@ -0,0 +1,137 @@
+   * @property {Array} keys
+   * @property {Function} create
+   * @property {Function} delete
+   * @property {Function} process
+   *
+   * Each method below returns a promise that resolves to an object,
+   * the shape of which is determined by config. If config has settings
+   * for `firstInterval` and `secondInterval` (as at time of writing),
+   * the shape of those objects would be `{ first, second }`.
+   */
+  return {
+    keys: keys.slice(),
+
+    /**
+     * Create verification reminder records for an account.
+     *
+     * @param {String} uid
+     * @returns {Promise} - Each property on the resolved object will be the number
+     *                      of elements added to that sorted set, i.e. the result of
+     *                      [`redis.zadd`](https://redis.io/commands/zadd).
+     */
+    async create (uid) {
+      try {
+        if (rolloutRate <= 1 && Math.random() < rolloutRate) {
+          const now = Date.now();
+          const result = await P.props(keys.reduce((result, key) => {
+            result[key] = redis.zadd(key, now, uid);
+            return result;
+          }, {}));
+          log.info('verificationReminders.create', { uid });
+          return result;
+        }
+      } catch (err) {
+        log.error('verificationReminders.create.error', { err, uid });
+        throw err;
+      }
+    },
+
+    /**
+     * Delete verification reminder records for an account.
+     *
+     * @param {String} uid
+     * @returns {Promise} - Each property on the resolved object will be the number of
+     *                      elements removed from that sorted set, i.e. the result of
+     *                      [`redis.zrem`](https://redis.io/commands/zrem).
+     */
+    async delete (uid) {
+      try {
+        const result = await P.props(keys.reduce((result, key) => {
+          result[key] = redis.zrem(key, uid);
+          return result;
+        }, {}));
+        log.info('verificationReminders.delete', { uid });
+        return result;
+      } catch (err) {
+        log.error('verificationReminders.delete.error', { err, uid });
+        throw err;
+      }
+    },
+
+    /**
+     * Read and remove all verification reminders that have
+     * ticked past the expiry intervals set in config.
+     *
+     * @returns {Promise} - Each property on the resolved object will be an array of uids that
+     *                      were found to have ticked past the relevant expiry interval, i.e.
+     *                      the result of [`redis.zrangebyscore`](https://redis.io/commands/zrangebyscore).
+     */
+    async process () {
+      try {
+        const now = Date.now();
+        return await P.props(keys.reduce((result, key, index) => {
+          const cutoff = now - intervals[index];
+          result[key] = redis.zrangebyscore(key, 0, cutoff);
+          setImmediate(async () => {
+            await result[key];
+            redis.zremrangebyscore(key, 0, cutoff);
```

Comment:
I know you are still working on the script but I was curious how it might handle failures, email bounce, spam, etc? Maybe it is ok to remove the entry and drop on failure?

Reply:
I'm not entirely happy with this logic either.

The main concern was how it would behave if two instances of the script are running at the same time, which is why I didn't want to wait until after the email has gone out to remove the data from Redis (because it could send the same email twice). But that is still possible here, albeit the time window in which it can happen is drastically shorter. Ideally we want an atomic `zrangebyscore+zremrangebyscore` operation, but I didn't find one of those yet.
Ideally we'd have something s

Reply:
Oh cool, turns out we can use `multi+exec` like we already do for session token updates:

https://github.com/antirez/redis/issues/117